polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala

```diff
@@ -137,66 +82,97 @@ class PythonInterpreter(val symbolTable: RuntimeSymbolTable) extends LanguageKer
-      val lastIsExpr = jep.getValue("isinstance(__polynote_last__, ast.Expr)", classOf[java.lang.Boolean])
-
-      val resultName = s"res$cell"
-      val maybeModifiedCode = if (lastIsExpr) {
-        val lastLine = jep.getValue("__polynote_last__.lineno", classOf[java.lang.Integer]) - 1
-        val (prevCode, lastCode) = code.linesWithSeparators.toSeq.splitAt(lastLine)
-
-        (prevCode ++ s"$resultName = (${lastCode.mkString})\n").mkString
-      } else code
-
-      jep.set("__polynote_code__", maybeModifiedCode)
-      jep.eval("exec(__polynote_code__, None, __polynote_locals__)\n")
-      jep.eval("globals().update(__polynote_locals__)")
-      val newDecls = jep.getValue("list(__polynote_locals__.keys())", classOf[java.util.List[String]]).asScala.toList
-
-      // TODO: We talked about potentially creating an `Out` map, inspired by ipython, instead of these resCell# vars
-      val maybeOutput = if (newDecls.last == resultName){
-        val hasLastValue = jep.getValue(s"$resultName != None", classOf[java.lang.Boolean])
-        if (hasLastValue) {
-          Option(Output("text/plain; rel=decl; lang=python", s"$resultName = ${jep.getValue(s"repr(${newDecls.last})", classOf[String])}"))
-        } else None
-      } else None
-
-      (getPyResults(newDecls), maybeOutput)
-
-    }.flatMap {
-      case (resultDict, maybeOutput) =>
-        val symbolPub = symbolTable.publishAll(resultDict.toList.filterNot(_._2 == null).map {
-          case (name, value) => symbolTable.RuntimeValue(name, value, Some(this), cell)
-        })
-
-        maybeOutput
-          .map(o => out.enqueue1(o))
-          .getOrElse(IO.unit)
-          .flatMap( _ => symbolPub)
-
-    }.flatMap {
-      _ => IO(stdout.flush())
+    runtimeContextView: runtimeContext.RuntimeContextView,
+    code: String
+  ): IO[(Stream[IO, Result], runtimeContext.RuntimeContextEntry)] = {
+    val prevCellId = Option(runtimeContextView.parentEntry.cellId)
+    if (code.trim().isEmpty) {
+      IO.pure((
+        Stream.empty,
+        runtimeContext.RuntimeContextEntry(cell, prevCellId, Map.empty, None, None)
+      ))
+    } else {
+
+      implicit val iShift: ContextShift[IO] = shift
+
+      def exec(stdout: PythonDisplayHook, globalDict: Map[String, Any]): IO[(Option[Output], Map[String, runtimeContext.globalInfo.RuntimeValue], Option[runtimeContext.globalInfo.RuntimeValue])] = {
+        withJep {
+
+          jep.set("__polynote_displayhook__", stdout)
+          jep.eval("import sys\n")
+          jep.eval("import ast\n")
+          jep.eval("sys.displayhook = __polynote_displayhook__.output\n")
+          jep.eval("sys.stdout = __polynote_displayhook__\n")
+          jep.eval("__polynote_locals__ = {}\n")
+
+          // TODO: maybe we need to add some more globals stuff here
+          // globals needs to be a dict but if we send a Map to jep it becomes a pyjmap, so we need to populate a dict
+          jep.eval("__polynote_globals__ = {}")
```

Comment:
Was thinking we might be able to wrap `runtimeContext` in a JVM object that has the python dict interface, and just `jep.set("__polynote_globals__")` to that. And it could lazily look up values as they're needed (and maybe do the inverse conversion that we're doing now, to put them nicely into python land)

Reply:
I was running into `globals needs to be a dict` error (or something like that). 

I guess it depends on how Python checks to see whether it is is a dictionary - if we could implement the interface then that would be awesome!